macros/string_mapping.sql

```diff
@@ -0,0 +1,69 @@
+{% macro string_mapping(table_name) -%}
+    {{ return(adapter.dispatch('string_mapping', 'dynamics_365_crm')(table_name)) }}
+{% endmacro %}
+
+{% macro default__string_mapping(table_name) %}
+    {{ config(enabled=var('dynamics_365_crm_using_' ~ table_name, True)) }}
+    {%- set columns = adapter.get_columns_in_relation(source('dynamics_365_crm', table_name)) -%}
+    {%- set attributes = dbt_utils.get_column_values(
+        table=source('dynamics_365_crm', 'stringmap'),
+        where="lower(objecttypecode) = '" ~ table_name ~ "'",
+        column='attributename') -%}
+
+    {%- set fields = [] -%}
+    {%- set non_pivot_fields = [] -%}
+    {%- for col in columns -%}
+        {%- if col.name | lower in attributes | map('lower') -%}
+            {%- do fields.append(col.name) -%}
+        {%- else -%}
+            {%- do non_pivot_fields.append(col.name) -%}
+        {%- endif -%}
+    {%- endfor -%}
+
+    with unpivoted as (
+        {{ dbt_utils.unpivot(
+            relation=source('dynamics_365_crm', table_name),
+            cast_to=dbt.type_string(),
+            exclude=non_pivot_fields,
+            field_name='fieldname',
+            value_name='fieldvalue'
+        ) }}
+
+    ), stringmaps as (
+
+        select
+            stringmapid,
+            cast(attributevalue as {{ dbt.type_string()}}) as attributevalue,
+            cast(attributename as {{ dbt.type_string()}}) as attributename,
+            cast(objecttypecode as {{ dbt.type_string()}}) as objecttypecode,
+            cast(value as {{ dbt.type_string()}}) as stringmap_value
+        from {{ source('dynamics_365_crm', 'stringmap')}}
+        where lower(objecttypecode) = {{ "'" ~ table_name ~ "'" }}
+        and not coalesce(_fivetran_deleted, false)
+
+    ), joined as (
+        select
+            unpivoted.*,
+            stringmaps.stringmap_value as fieldvalue_name
+        from unpivoted
+        left join stringmaps
+        on lower(unpivoted.fieldname) = lower(stringmaps.attributename)
+        and lower(unpivoted.fieldvalue) = lower(stringmaps.attributevalue)
+
+    ), repivoted as (
+        select
+            {% for field in non_pivot_fields %}
+                {{ field }},
+            {% endfor %}
+            {% for field in fields %}
+                max(case when lower(fieldname) = lower('{{ field }}') then fieldvalue else null end) as {{ field }},
+                max(case when lower(fieldname) = lower('{{ field }}') then fieldvalue_name else null end) as {{ field }}_name{{ ',' if not loop.last }}
```

Comment:
I'm not sure how I feel about appending `_name` to the string mapped value field. I mention because I saw in some test data that some of non string mapped had the same `_name` naming convention. It could be confusing for customers to see a common naming convention and not be sure which are generated and which aren't. My instinct is saying we should make these clearly distinguishable from the other non mapped fields.

Let's sync with Sadie on this to align on a naming convention. I will say `_name` could still work, but I would like to have a discussion around it to make sure it's the best option.